ollama/_utils.py

```diff
@@ -0,0 +1,212 @@
+from __future__ import annotations
+
+from typing import Any, Callable, List, Mapping, Optional, Union, get_args, get_origin
+from ollama._types import Tool
+from collections.abc import Sequence, Set
+from typing import Dict, Set as TypeSet
+import sys
+
+# Type compatibility layer for Union and UnionType
+if sys.version_info >= (3, 10):
+  from types import UnionType
+
+  def is_union(tp: Any) -> bool:
+    return get_origin(tp) in (Union, UnionType)
+else:
+
+  def is_union(tp: Any) -> bool:
+    return get_origin(tp) is Union
+
+
+# Map both the type and the type reference to the same JSON type
+TYPE_MAP = {
+  # Basic types
+  int: 'integer',
+  'int': 'integer',
+  str: 'string',
+  'str': 'string',
+  float: 'number',
+  'float': 'number',
+  bool: 'boolean',
+  'bool': 'boolean',
+  type(None): 'null',
+  None: 'null',
+  'None': 'null',
+  # Collection types
+  list: 'array',
+  'list': 'array',
+  List: 'array',
+  'List': 'array',
+  Sequence: 'array',
+  'Sequence': 'array',
+  tuple: 'array',
+  'tuple': 'array',
+  set: 'array',
+  'set': 'array',
+  Set: 'array',
+  TypeSet: 'array',
+  'Set': 'array',
+  # Mapping types
+  dict: 'object',
+  'dict': 'object',
+  Dict: 'object',
+  'Dict': 'object',
+  Mapping: 'object',
+  'Mapping': 'object',
+  Any: 'string',
+  'Any': 'string',
+}
+
+
+def _get_json_type(python_type: Any) -> str | List[str]:
+  # Handle Optional types (Union[type, None] and type | None)
+  if is_union(python_type):
+    args = get_args(python_type)
+    # Filter out None/NoneType from union args
+    if non_none_args := [arg for arg in args if arg not in (None, type(None))]:
+      if len(non_none_args) == 1:
+        return _get_json_type(non_none_args[0])
+      # For multiple return types (e.g., int | str | None), return stringified array of types -> "['integer', 'string', 'null']"
+      return str([_get_json_type(arg) for arg in non_none_args]).replace(' ', '')
```

Comment:
added a stringify here to not have to change the types on ollama/ollama repo. Function calling working as expected with change 😁